### PR TITLE
Allow filtering by category

### PR DESCRIPTION
**Pull request by Krenair**
This doesn't display anything in the queue filter GUI yet. That'll be a future commit.

Bug: T72723

### CODE REVIEW

File: huggle/hugglequeuefilter.hpp [hunk 1]
```diff
@@ -113,6 +119,9 @@ namespace Huggle
             HuggleQueueFilterMatch Self;
             HuggleQueueFilterMatch UserSpace;
             HuggleQueueFilterMatch TalkPage;
+            QString Type;
```
Comment (benapetr) [inline huggle/hugglequeuefilter.hpp]: Looked more into this I don't really understand what is point of all these 3 variables? Filter is only 1 for whole queue, so list of categories you want to exclude, or require should be defined in each filter separately. What are these 3 variables for? And how do you store a list of categories that user wants to filter now?

Comment (Krenair) [inline huggle/hugglequeuefilter.hpp]: The variables are set in the DefaultConfig
We could allow filtering for one of (or all of) a set of categories if you want

File: huggle/wikiedit.cpp [hunk 2]
```diff
@@ -106,6 +106,24 @@ bool WikiEdit::FinalizePostProcessing()
         return false;
     }
 
+    if (this->qCategories != nullptr && this->qCategories->IsProcessed())
+    {
+        if (this->qCategories->IsFailed())
+        {
+            Syslog::HuggleLogs->ErrorLog("Unable to fetch categories for page " + this->Page->PageName + ": " + this->qCategories->GetFailureReason());
```
Comment (Krenair) [inline huggle/wikiedit.cpp]: I think this was just following the existing pattern for other post-process queries

**Comment by Krenair** [comment]
I have tested this with the default "Candidates for speedy deletion" filter. I created my own in https://en.wikipedia.org/w/index.php?title=User%3AKrenair%2Fhuggle3.css&type=revision&diff=773914330&oldid=773805466

File: huggle/wikiedit.cpp (hunk 2 above)
Comment (benapetr) [inline huggle/wikiedit.cpp]: Yes they all need to be made translatable, there is a GCI task for that

**Comment by benapetr** [comment]
OK,

But I think of a little different design here, basically right now we can have only 1 filter active in a moment. The way how you implemented this, it's not possible to filter multiple categories in each filter. Imagine someone wanted to create a filter that ignores talk pages, bots and registered users and wants to see only pages in category "United States" which also are not in category "Presidents" and "Politicians". This is not possible with your design.

So what I propose (I still don't understand what is meaning of these 3 variables you created, but whatever) - add this filter in same way as tag filter was implemented. Add 2 string lists "IgnoreCats" and "RequireCats" and let user define a list of categories for each (separated either by comma or newline, I don't know what is better).

Then the filter would just look if wikiedit has all "RequireCats" and doesn't have any of "IgnoreCats".

I hope it's clear what I mean :)

**Comment by benapetr** [comment]
And in your configuration file huggle3.css it would look like this:

```
queues:
    User defined queue #9:
        filter-ignored:exclude
        filter-bots:exclude
        filter-assisted:exclude
        filter-ip:ignore
        filter-minor:ignore
        filter-new-pages:ignore
        filter-me:exclude
        filter-users:ignore
        nsfilter-user:ignore
        filter-talk:exclude
        ignored-tags:mobileaedit,asd
        required-tags:
        required-cats:United States
        ignored-cats:Politicians, Presidents
```

I omitted these 3 variables you added because I just don't understand their purpose. However this is I believe what users on feedback page wanted to have

**Comment by Krenair** [comment]
Sure, I was just trying to stick to the existing config. I'll change it around to work like that.

File: huggle/wikiedit.cpp (hunk 2 above)
Comment (Krenair) [inline huggle/wikiedit.cpp]: So I'll leave this as is and it can be followed up with the GCI task?

**Comment by Krenair** [comment]
By the way, why does a lot of the default config appear to not be used?
 https://phabricator.wikimedia.org/diffusion/GHUG/browse/master/huggle/Resources/DefaultConfig.txt$486,488-489,491-494,499,500,512,516,519-521

**Comment by benapetr** [comment]
Because huggle3 is a rewrite of older version of huggle. Some of these were not implemented yet.

**Comment by benapetr** [comment]
+ Some of these we probably don't want to implement. I will have a look how categories worked before, but if we have a chance to do it better way this time, I think we should use it :))

**Comment by Krenair** [comment]
@benapetr: I don't think the GUI commit got merged?

**Comment by Krenair** [comment]
nevermind, it got merged, had some issues with git. the way github merges commits together is a bit strange

**Pull request merged**